File: tefloader/net/Il2CppApi/Array.cs
// /*******************************************************************************
//  * tefloader - Array.cs
//  * Copyright (C) 2026 eternalfuture-e38299
//  *
//  * This program is free software: you can redistribute it and/or modify
//  * it under the terms of the GNU Affero General Public License as published by
//  * the Free Software Foundation, either version 3 of the License, or
//  * (at your option) any later version.
//  *
//  * This program is distributed in the hope that it will be useful,
//  * but WITHOUT ANY WARRANTY; without even the implied warranty of
//  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
//  * GNU Affero General Public License for more details.
//  *
//  * You should have received a copy of the GNU Affero General Public License
//  * along with this program. If not, see <https://www.gnu.org/licenses/>.
//  *
//  * Author: eternalfuture-e38299
//  * GitHub: https://github.com/eternalfuture-e38299
//  * Created: $[InvalidReference]
//  *******************************************************************************/

using System.Runtime.InteropServices;

namespace tefloader.Il2CppApi;

public static class Array
{
    // ==================== 基础 API ====================

    // 模拟 il2cpp_array_new - 创建新数组（保持不变）
    public static IntPtr il2cpp_array_new(IntPtr elementTypeInfo, int length)
    {
        if (elementTypeInfo == IntPtr.Zero || length <= 0)
            return IntPtr.Zero;

        try
        {
            var typeHandle = GCHandle.FromIntPtr(elementTypeInfo);
            var elementType = (Type)typeHandle.Target;

            // 创建托管数组
            var managedArray = System.Array.CreateInstance(elementType, length);

            var arrayHandle = GCHandle.Alloc(managedArray);
            return GCHandle.ToIntPtr(arrayHandle);
        }
        catch (Exception ex)
        {
            Logger.Error($"Failed to create array: {ex.Message}");
            return IntPtr.Zero;
        }
    }

    // 模拟 il2cpp_array_element_size - 获取元素大小（保持不变）
    public static int il2cpp_array_element_size(IntPtr arrayPtr)
    {
        if (arrayPtr == IntPtr.Zero)
            return 0;

        try
        {
            var arrayHandle = GCHandle.FromIntPtr(arrayPtr);
            var array = (System.Array)arrayHandle.Target;

            var elementType = array.GetType().GetElementType();
            return elementType == null ? 0 : Marshal.SizeOf(elementType);
        }
        catch (Exception ex)
        {
            Logger.Error($"Failed to get element size: {ex.Message}");
            return 0;
        }
    }

    // 模拟 il2cpp_array_length - 获取数组总长度（保持不变）
    public static int il2cpp_array_length(IntPtr arrayPtr)
    {
        if (arrayPtr == IntPtr.Zero)
            return 0;

        try
        {
            var arrayHandle = GCHandle.FromIntPtr(arrayPtr);
            var array = (System.Array)arrayHandle.Target;

            return array.Length;
        }
        catch (Exception ex)
        {
            Logger.Error($"Failed to get array length: {ex.Message}");
            return 0;
        }
    }

    // il2cpp_array_resize - 扩容数组：创建新数组并复制旧数据，返回新数组句柄（失败返回 IntPtr.Zero）
    public static IntPtr il2cpp_array_resize(IntPtr arrayPtr, int newSize)
    {
        if (arrayPtr == IntPtr.Zero || newSize <= 0)
            return IntPtr.Zero;

        try
        {
            var arrayHandle = GCHandle.FromIntPtr(arrayPtr);
            var array = (System.Array)arrayHandle.Target;

            if (array == null || newSize <= array.Length)
                return IntPtr.Zero;

            var elementType = array.GetType().GetElementType();
            if (elementType == null)
                return IntPtr.Zero;

            var newArray = System.Array.CreateInstance(elementType, newSize);
            System.Array.Copy(array, newArray, array.Length);

            var newHandle = GCHandle.Alloc(newArray);
            return GCHandle.ToIntPtr(newHandle);
        }
        catch (Exception ex)
        {
            Logger.Error($"Failed to resize array: {ex.Message}");
            return IntPtr.Zero;
        }
    }

    // ==================== 辅助方法：多维索引转换 ====================

    // 将一维索引转换为多维索引（行主序，C语言风格）
    private static int[] GetMultiDimensionalIndices(System.Array array, int flatIndex)
    {
        if (array.Rank == 1)
            return [flatIndex];

        var indices = new int[array.Rank];
        var remaining = flatIndex;

        // 从最后一维开始计算（行主序）
        for (var i = array.Rank - 1; i >= 0; i--)
        {
            var dimSize = array.GetLength(i);
            indices[i] = remaining % dimSize;
            remaining /= dimSize;
        }

        return indices;
    }

    // ==================== 元素操作 ====================

    // 额外功能：il2cpp_array_at - 获取数组元素（支持多维数组的一维索引访问）
    public static unsafe bool il2cpp_array_at(IntPtr arrayPtr, int index, void* outValue)
    {
        if (arrayPtr == IntPtr.Zero || outValue == null || index < 0)
            return false;

        try
        {
            var arrayHandle = GCHandle.FromIntPtr(arrayPtr);
            var array = (System.Array)arrayHandle.Target;

            if (index >= array.Length)
                return false;

            // 如果是多维数组，将一维索引转换为多维索引
            object value;
            if (array.Rank > 1)
            {
                var indices = GetMultiDimensionalIndices(array, index);
                value = array.GetValue(indices);
            }
            else
            {
                value = array.GetValue(index);
            }

            if (value == null)
                return false;

            var elementType = array.GetType().GetElementType();

            if (elementType!.IsValueType)
                // 值类型直接复制
            {
                Marshal.StructureToPtr(value, (IntPtr)outValue, false);
            }
            else
            {
                // 引用类型，分配 GCHandle
                var handle = GCHandle.Alloc(value);
                *(IntPtr*)outValue = GCHandle.ToIntPtr(handle);
            }

            return true;
        }
        catch (Exception ex)
        {
            Logger.Error($"Failed to get array element: {ex.Message}");
            return false;
        }
    }

    // 额外功能：il2cpp_array_set - 设置数组元素（支持多维数组的一维索引访问）
    public static unsafe bool il2cpp_array_set(IntPtr arrayPtr, int index, void* value)
    {
        if (arrayPtr == IntPtr.Zero || value == null || index < 0)
            return false;

        try
        {
            var arrayHandle = GCHandle.FromIntPtr(arrayPtr);
            var array = (System.Array)arrayHandle.Target;

            if (index >= array.Length)
                return false;

            var elementType = array.GetType().GetElementType();
            object? newValue;

            if (elementType!.IsValueType)
            {
                // 值类型从指针读取
                newValue = Marshal.PtrToStructure((IntPtr)value, elementType);
            }
            else
            {
                // 引用类型从 GCHandle 获取
                var objPtr = *(IntPtr*)value;
                if (objPtr == IntPtr.Zero)
                    newValue = null;
                else
                    newValue = Utils.PtrToObject(objPtr);
            }

            // 如果是多维数组，将一维索引转换为多维索引
            if (array.Rank > 1)
            {
                var indices = GetMultiDimensionalIndices(array, index);
                array.SetValue(newValue, indices);
            }
            else
            {
                array.SetValue(newValue, index);
            }

            return true;
        }
        catch (Exception ex)
        {
            Logger.Error($"Failed to set array element: {ex.Message}");
            return false;
        }
    }

    // 额外功能：il2cpp_array_fill - 填充数组（支持多维数组）
    public static unsafe bool il2cpp_array_fill(IntPtr arrayPtr, void* value)
    {
        if (arrayPtr == IntPtr.Zero || value == null)
            return false;

        try
        {
            var arrayHandle = GCHandle.FromIntPtr(arrayPtr);
            var array = (System.Array)arrayHandle.Target;

            var elementType = array.GetType().GetElementType();

            if (elementType!.IsValueType)
            {
                // 值类型：读取一次值，然后填充所有元素
                var fillValue = Marshal.PtrToStructure((IntPtr)value, elementType);
                for (var i = 0; i < array.Length; i++)
                    if (array.Rank > 1)
                    {
                        var indices = GetMultiDimensionalIndices(array, i);
                        array.SetValue(fillValue, indices);
                    }
                    else
                    {
                        array.SetValue(fillValue, i);
                    }
            }
            else
            {
                // 引用类型：从 GCHandle 获取对象
                var objPtr = *(IntPtr*)value;
                var fillValue = objPtr == IntPtr.Zero ? null : Utils.PtrToObject(objPtr);
                for (var i = 0; i < array.Length; i++)
                    if (array.Rank > 1)
                    {
                        var indices = GetMultiDimensionalIndices(array, i);
                        array.SetValue(fillValue, indices);
                    }
                    else
                    {
                        array.SetValue(fillValue, i);
                    }
            }

            return true;
        }
        catch (Exception ex)
        {
            Logger.Error($"Failed to fill array: {ex.Message}");
            return false;
        }
    }

    // ==================== C 数组与托管数组之间的复制 ====================

    // 额外功能：il2cpp_array_copy_from_c - 从 C 数组复制到托管数组（支持引用类型）
    public static unsafe bool il2cpp_array_copy_from_c(IntPtr destArrayPtr, void* src, int count)
    {
        if (destArrayPtr == IntPtr.Zero || src == null || count <= 0)
            return false;

        try
        {
            var arrayHandle = GCHandle.FromIntPtr(destArrayPtr);
            var destArray = (System.Array)arrayHandle.Target;

            if (destArray == null)
                return false;

            if (count > destArray.Length)
                count = destArray.Length;

            var elementType = destArray.GetType().GetElementType();

            // 检查元素类型
            if (elementType == null)
                return false;

            // 获取元素大小（用于值类型）
            int elementSize;
            var isValueType = elementType.IsValueType;

            if (isValueType)
            {
                elementSize = Marshal.SizeOf(elementType);

                // 值类型：直接从内存复制
                for (var i = 0; i < count; i++)
                {
                    var elementPtr = (IntPtr)src + i * elementSize;
                    var value = Marshal.PtrToStructure(elementPtr, elementType);

                    if (destArray.Rank > 1)
                    {
                        var indices = GetMultiDimensionalIndices(destArray, i);
                        destArray.SetValue(value, indices);
                    }
                    else
                    {
                        destArray.SetValue(value, i);
                    }
                }
            }
            else
            {
                // 引用类型：从 GCHandle 指针数组复制
                // 注意：src 指向的是 GCHandle 指针数组，每个元素是 IntPtr
                for (var i = 0; i < count; i++)
                {
                    // 读取 GCHandle 指针
                    var handlePtr = *(IntPtr*)((byte*)src + i * IntPtr.Size);
                    object? value;

                    if (handlePtr == IntPtr.Zero)
                        value = null;
                    else
                        try
                        {
                            // 从 GCHandle 获取对象
                            var handle = GCHandle.FromIntPtr(handlePtr);
                            value = handle.Target;
                        }
                        catch
                        {
                            // 如果无法从 GCHandle 获取，尝试直接转换指针为对象
                            value = Utils.PtrToObject(handlePtr);
                        }

                    if (destArray.Rank > 1)
                    {
                        var indices = GetMultiDimensionalIndices(destArray, i);
                        destArray.SetValue(value, indices);
                    }
                    else
                    {
                        destArray.SetValue(value, i);
                    }
                }
            }

            return true;
        }
        catch (Exception ex)
        {
            Logger.Error($"Failed to copy from C array: {ex.Message}");
            return false;
        }
    }

// 额外功能：il2cpp_array_copy_to_c - 从托管数组复制到 C 数组（支持引用类型）
    public static unsafe bool il2cpp_array_copy_to_c(void* dest, IntPtr srcArrayPtr, int count)
    {
        if (dest == null || srcArrayPtr == IntPtr.Zero || count <= 0)
            return false;

        try
        {
            var arrayHandle = GCHandle.FromIntPtr(srcArrayPtr);
            var srcArray = (System.Array)arrayHandle.Target;

            if (srcArray == null)
                return false;

            if (count > srcArray.Length)
                count = srcArray.Length;

            var elementType = srcArray.GetType().GetElementType();

            if (elementType == null)
                return false;

            var isValueType = elementType.IsValueType;

            if (isValueType)
            {
                // 值类型：直接复制到内存
                var elementSize = Marshal.SizeOf(elementType);

                for (var i = 0; i < count; i++)
                {
                    object? value;
                    if (srcArray.Rank > 1)
                    {
                        var indices = GetMultiDimensionalIndices(srcArray, i);
                        value = srcArray.GetValue(indices);
                    }
                    else
                    {
                        value = srcArray.GetValue(i);
                    }

                    var elementPtr = (IntPtr)dest + i * elementSize;

                    if (value != null)
                        Marshal.StructureToPtr(value, elementPtr, false);
                    else
                        // 如果是 null，清零
                        *(long*)elementPtr = 0;
                }
            }
            else
            {
                // 引用类型：存储 GCHandle 指针到 C 数组
                for (var i = 0; i < count; i++)
                {
                    object? value;
                    if (srcArray.Rank > 1)
                    {
                        var indices = GetMultiDimensionalIndices(srcArray, i);
                        value = srcArray.GetValue(indices);
                    }
                    else
                    {
                        value = srcArray.GetValue(i);
                    }

                    var destPtr = (IntPtr*)dest + i;

                    if (value == null)
                    {
                        *destPtr = IntPtr.Zero;
                    }
                    else
                    {
                        // 分配 GCHandle 并返回指针
                        var handle = GCHandle.Alloc(value);
                        *destPtr = GCHandle.ToIntPtr(handle);
                    }
                }
            }

            return true;
        }
        catch (Exception ex)
        {
            Logger.Error($"Failed to copy to C array: {ex.Message}");
            return false;
        }
    }

    // il2cpp_array_copy - 直接从源数组复制到目标数组（高性能版本）
    public static bool il2cpp_array_copy(IntPtr destArrayPtr, IntPtr srcArrayPtr, int count)
    {
        if (destArrayPtr == IntPtr.Zero || srcArrayPtr == IntPtr.Zero)
            return false;

        try
        {
            var destHandle = GCHandle.FromIntPtr(destArrayPtr);
            var srcHandle = GCHandle.FromIntPtr(srcArrayPtr);

            var destArray = (System.Array)destHandle.Target;
            var srcArray = (System.Array)srcHandle.Target;

            if (destArray == null || srcArray == null)
                return false;

            // 自动适应
            count = count <= 0
                ? Math.Min(destArray.Length, srcArray.Length)
                : Math.Min(count, Math.Min(destArray.Length, srcArray.Length));

            if (count <= 0)
                return true;

            var destElementType = destArray.GetType().GetElementType();
            var srcElementType = srcArray.GetType().GetElementType();

            // 如果类型相同且是一维数组，使用 Array.Copy（最快）
            if (destElementType == srcElementType && destArray.Rank == 1 && srcArray.Rank == 1)
            {
                System.Array.Copy(srcArray, 0, destArray, 0, count);
                return true;
            }

            // 如果类型相同但可能是多维数组，使用 Array.Copy 也能处理
            if (destElementType == srcElementType)
            {
                // Array.Copy 支持多维数组
                System.Array.Copy(srcArray, 0, destArray, 0, count);
                return true;
            }

            // 类型不同，需要逐元素转换（保持原有逻辑）
            Logger.Warning($"Copying between different element types: {srcElementType} -> {destElementType}");

            for (var i = 0; i < count; i++)
            {
                object? value;

                // 从源数组读取
                if (srcArray.Rank > 1)
                {
                    var indices = GetMultiDimensionalIndices(srcArray, i);
                    value = srcArray.GetValue(indices);
                }
                else
                {
                    value = srcArray.GetValue(i);
                }

                // 类型转换
                if (value != null && value.GetType() != destElementType)
                    try
                    {
                        value = Convert.ChangeType(value, destElementType!);
                    }
                    catch
                    {
                        Logger.Warning($"Cannot convert {value.GetType()} to {destElementType}");
                    }

                // 写入目标数组
                if (destArray.Rank > 1)
                {
                    var indices = GetMultiDimensionalIndices(destArray, i);
                    destArray.SetValue(value, indices);
                }
                else
                {
                    destArray.SetValue(value, i);
                }
            }

            return true;
        }
        catch (Exception ex)
        {
            Logger.Error($"Failed to copy array: {ex.Message}");
            return false;
        }
    }
}

File: src/patchlib/android/struct/array.c
/*******************************************************************************
 * tefkernel - array
 * Copyright (C) 2025 eternalfuture-e38299
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 *
 * Author: eternalfuture-e38299
 * GitHub: https://github.com/eternalfuture-e38299
 * Created: 2025/12/27
 * Modified: 2026/08/04 - Removed element size limit to support structs
 *******************************************************************************/
#include "patchlib/struct/array.h"

#include <stdint.h>
#include <string.h>
#include "internal/log.h"
#include "../../il2cpp_api.h"

// 移除最大元素大小限制，支持任意大小的结构体
// #define MAX_ELEMENT_SIZE sizeof(void*)  // 已移除

// ReSharper disable once CppUseInternalLinkage
typedef struct il2cpp_array_t {
    // ReSharper disable once CppDeclaratorNeverUsed
    void *m_class;
    // ReSharper disable once CppDeclaratorNeverUsed
    void *m_monitor;
    // ReSharper disable once CppDeclaratorNeverUsed
    void *m_bounds;
    // ReSharper disable once CppDeclaratorNeverUsed
    uint32_t m_length;
    // T *m_values;
} il2cpp_array_t;


// 获取数组数据起始位置
#define ARRAY_DATA_START(array) ((char*)(array) + sizeof(il2cpp_array_t))

// 获取数组元素大小（通过字节长度和长度计算）
static size_t get_array_element_size(patch_handle_t array) {
    const uint32_t byte_length = il2cpp_array_get_byte_length(array);
    const uint32_t length = il2cpp_array_length(array);

    if (length == 0) {
        return 0;
    }

    const size_t element_size = byte_length / length;

    // 调试日志：输出元素大小
    TEKLOG_DEBUG("Array element size: %zu bytes (length=%u, byte_length=%u)",
                 element_size, length, byte_length);

    return element_size;
}

// 检查元素大小是否合法（移除最大限制）
static bool is_element_size_valid(size_t element_size) {
    if (element_size == 0) {
        TEKLOG_ERROR("Invalid element size: 0");
        return false;
    }
    // 只检查是否过大（防止溢出）
    if (element_size > 1024 * 1024) { // 1MB 上限保护
        TEKLOG_ERROR("Element size too large: %zu bytes (max: 1MB)", element_size);
        return false;
    }
    return true;
}

bool patchlib_array_at(patch_handle_t array, const size_t index, void* out_value) {
    if (!patchlib_is_valid(array)) {
        TEKLOG_ERROR("Invalid array handle");
        return false;
    }

    if (out_value == NULL) {
        TEKLOG_ERROR("Output value pointer is NULL");
        return false;
    }

    const uint32_t length = il2cpp_array_length(array);
    if (index >= length) {
        TEKLOG_ERROR("Index %zu out of bounds (length=%u)", index, length);
        return false;
    }

    const size_t element_size = get_array_element_size(array);
    if (!is_element_size_valid(element_size)) {
        return false;
    }

    const char* data_start = ARRAY_DATA_START(array);
    const char* element_ptr = data_start + (index * element_size);

    memcpy(out_value, element_ptr, element_size);
    return true;
}

bool patchlib_array_set(patch_handle_t array, const size_t index, void* new_value) {
    if (!patchlib_is_valid(array)) {
        TEKLOG_ERROR("Invalid array handle");
        return false;
    }

    if (!new_value) {
        TEKLOG_ERROR("New value pointer is NULL");
        return false;
    }

    const uint32_t length = il2cpp_array_length(array);
    if (index >= length) {
        TEKLOG_ERROR("Index %zu out of bounds (length=%u)", index, length);
        return false;
    }

    const size_t element_size = get_array_element_size(array);
    if (!is_element_size_valid(element_size)) {
        return false;
    }

    char* data_start = ARRAY_DATA_START(array);
    char* element_ptr = data_start + (index * element_size);

    memcpy(element_ptr, new_value, element_size);

    TEKLOG_DEBUG("Set array[%zu]: element_size=%zu", index, element_size);

    return true;
}

bool patchlib_array_fill(patch_handle_t array, void* value) {
    if (!patchlib_is_valid(array)) {
        TEKLOG_ERROR("Invalid array handle");
        return false;
    }

    if (!value) {
        TEKLOG_ERROR("Value pointer is NULL");
        return false;
    }

    const uint32_t length = il2cpp_array_length(array);
    const size_t element_size = get_array_element_size(array);

    if (!is_element_size_valid(element_size)) {
        return false;
    }

    char* data_start = ARRAY_DATA_START(array);

    // 优化：根据元素大小选择最优的填充方式
    if (element_size == sizeof(uint8_t)) {
        const uint8_t byte_value = *(uint8_t*)value;
        memset(data_start, byte_value, length);
    } else if (element_size == sizeof(uint64_t) && *(uint64_t*)value == 0 || element_size == sizeof(uint32_t) && *(uint32_t*)value == 0) {
        memset(data_start, 0, length * element_size);
    } else {
        for (size_t i = 0; i < length; ++i) {
            char* element_ptr = data_start + i * element_size;
            memcpy(element_ptr, value, element_size);
        }
    }

    return true;
}

bool patchlib_array_copy_from_c(patch_handle_t dest, const void* src, const size_t count) {
    if (!patchlib_is_valid(dest)) {
        TEKLOG_ERROR("Invalid destination array handle");
        return false;
    }

    if (!src) {
        TEKLOG_ERROR("Source C array is NULL");
        return false;
    }

    const uint32_t dest_len = il2cpp_array_length(dest);
    if (count > dest_len) {
        TEKLOG_ERROR("Count %zu exceeds destination length %u", count, dest_len);
        return false;
    }

    const size_t element_size = get_array_element_size(dest);
    if (!is_element_size_valid(element_size)) {
        return false;
    }

    char* dest_data = ARRAY_DATA_START(dest);
    memcpy(dest_data, src, count * element_size);

    TEKLOG_DEBUG("Copied %zu elements (size=%zu) to array", count, element_size);

    return true;
}

bool patchlib_array_copy_to_c(void* dest, patch_handle_t src, const size_t count) {
    if (!dest) {
        TEKLOG_ERROR("Destination C array is NULL");
        return false;
    }

    if (!patchlib_is_valid(src)) {
        TEKLOG_ERROR("Invalid source array handle");
        return false;
    }

    const uint32_t src_len = il2cpp_array_length(src);
    if (count > src_len) {
        TEKLOG_ERROR("Count %zu exceeds source length %u", count, src_len);
        return false;
    }

    const size_t element_size = get_array_element_size(src);
    if (!is_element_size_valid(element_size)) {
        return false;
    }

    const char* src_data = ARRAY_DATA_START(src);
    memcpy(dest, src_data, count * element_size);

    TEKLOG_DEBUG("Copied %zu elements (size=%zu) from array", count, element_size);

    return true;
}

bool patchlib_array_copy(patch_handle_t dest, patch_handle_t src, const size_t count) {
    // 参数验证
    if (!patchlib_is_valid(dest)) {
        TEKLOG_ERROR("Invalid destination array handle");
        return false;
    }

    if (!patchlib_is_valid(src)) {
        TEKLOG_ERROR("Invalid source array handle");
        return false;
    }

    if (count == 0) {
        TEKLOG_DEBUG("Copy count is 0, skipping");
        return true;  // 空操作
    }

    // 获取数组长度
    const uint32_t dest_len = il2cpp_array_length(dest);
    const uint32_t src_len = il2cpp_array_length(src);

    // 边界检查
    if (count > dest_len) {
        TEKLOG_ERROR("Count %zu exceeds destination array length %u", count, dest_len);
        return false;
    }

    if (count > src_len) {
        TEKLOG_ERROR("Count %zu exceeds source array length %u", count, src_len);
        return false;
    }

    // 获取元素大小
    const size_t dest_element_size = get_array_element_size(dest);
    const size_t src_element_size = get_array_element_size(src);

    if (!is_element_size_valid(dest_element_size) || !is_element_size_valid(src_element_size)) {
        return false;
    }

    // 检查元素大小是否一致
    if (dest_element_size != src_element_size) {
        TEKLOG_ERROR("Element size mismatch: dest=%zu, src=%zu",
                     dest_element_size, src_element_size);
        return false;
    }

    // 计算数据指针
    char* dest_data = ARRAY_DATA_START(dest);
    const char* src_data = ARRAY_DATA_START(src);

    // 复制数据
    const size_t total_bytes = count * dest_element_size;
    memcpy(dest_data, src_data, total_bytes);

    TEKLOG_DEBUG("Copied %zu elements (size=%zu, total=%zu bytes) from src to dest",
                 count, dest_element_size, total_bytes);

    return true;
}